plugins/woocommerce/includes/class-wc-session-handler.php

```diff
@@ -94,11 +94,26 @@ public function init_session_cookie() {
 			$this->_session_expiration = $cookie[1];
 			$this->_session_expiring   = $cookie[2];
 			$this->_has_cookie         = true;
-			$this->_data               = $this->get_session_data();
+
+			/**
+			 * Allows filtering of the session data retrieved from storage during session cookie initialization.
+			 *
+			 * @param array $session_data The session data as retrieved from storage.
+			 *
+			 * @since 9.9.0
+			 */
+			$this->_data = apply_filters( 'woocommerce_restored_session_data', $this->get_session_data() );
 
 			if ( ! $this->is_session_cookie_valid() ) {
 				$this->destroy_session();
 				$this->set_session_expiration();
+			} elseif ( empty( $this->_data ) && ! isset( WC()->cart ) ) {
```

Comment:
Possibly it's a silly question, but are we sure this is not affecting logged-in users? Like triggering a new session creation?

Reply:
Not a silly question at all.  And the answer is yes and no.

Because the user is logged in, `WC_Session_Handler->has_session()` returns `true`, which means that `WC_Session_Hander->save_data()` will save back to the DB if the session is "dirty" / has changes.  However, the session cookie is still removed (marked expired).  The cookie won't get added back to the browser unless something calls `WC_Session_Handler::set_customer_session_cookie( true )` - which is mainly only called in core  when the cart isn't empty.